jax/_src/lax/control_flow.py

```diff
@@ -326,8 +330,9 @@ def _create_jaxpr(init_val):
                       body_nconsts=len(body_consts), body_jaxpr=body_jaxpr)
   return tree_unflatten(body_tree, outs)
 
-def _while_loop_abstract_eval(*args, **kwargs):
-  return _map(raise_to_shaped, kwargs["body_jaxpr"].out_avals)
+def _while_loop_abstract_eval(*args, body_jaxpr, **kwargs):
+  del args, kwargs
+  return _map(raise_to_shaped, body_jaxpr.out_avals), body_jaxpr.effects
```

Comment:
Can the cond_jaxpr have effects? The code above suggests yes. Then the effect of applying a while_loop isn't just the effect of the body.

I think for this PR we should just raise NotImplementedError in the `while_loop` traceable if the jaxprs formed have any effects at all. Then let's add support for effects to each control flow primitive one by one.

Reply:
Done.